syntax = "proto3";

package ghostpacer.frame;

option go_package = "github.com/Ghost-Pacer/protocols/ghostpacer/frame/go/frameproto";

message Vec3 {
  double x = 1;
  double y = 2;
  double z = 3;
}

message Mat3 {
  Vec3 row1 = 1;
  Vec3 row2 = 2;
  Vec3 row3 = 3;
}

message Avatar {
  reserved 2, 5;

  Vec3 avatarPosition = 1;
  Vec3 avatarToVec = 6;
  Vec3 avatarUpVec = 7;
  double animationPhase = 3;

  // Camera vectors should not be included in final protobuf
  Vec3 cameraPosition = 4;
  Vec3 cameraToVec = 8;
  Vec3 cameraUpVec = 9;

  uint32 modelIndex = 10;
}

message Menu {
  repeated string entries = 1;
  uint32 activeEntryIndex = 2;
}

message Banner {
  enum Position {
    TOP = 0;
    CENTER = 1;
  }
  repeated string lines = 1;
  Position position = 2;

}

message Legend {
  bool upActive = 1;
  bool selectActive = 2;
  bool downActive = 3;
}

message Frame {
  Avatar avatar1 = 1;
  Avatar avatar2 = 2;
  Legend legend = 3;
  Menu menu = 4;
  Banner banner = 5;

  Vec3 cameraPosition = 6;
  Vec3 cameraToVec = 7;
  Vec3 cameraUpVec = 8;
}
